R/pkratio-datamapping.R

```diff
@@ -10,35 +10,92 @@ PKRatioDataMapping <- R6::R6Class(
+
+      colorGroupingList <- self$get_grouping_list(colorGrouping,colorGroupingDataFrame)
+      sizeGroupingList  <- self$get_grouping_list(sizeGrouping,sizeGroupingDataFrame)
+      shapeGroupingList <- self$get_grouping_list(shapeGrouping,shapeGroupingDataFrame)
+
+      self$colorGrouping <- colorGroupingList$selfGrouping
+      self$sizeGrouping  <- sizeGroupingList$selfGrouping
+      self$shapeGrouping <- shapeGroupingList$selfGrouping
+
+      self$colorGroupingDataFrame <- colorGroupingList$selfGroupingDataFrame
+      self$sizeGroupingDataFrame  <- sizeGroupingList$selfGroupingDataFrame
+      self$shapeGroupingDataFrame <- shapeGroupingList$selfGroupingDataFrame
+
+      self$colorLegendTitle = colorLegendTitle
+      self$sizeLegendTitle = sizeLegendTitle
+      self$shapeLegendTitle = shapeLegendTitle
 
       super$initialize(x, y)
     },
 
+
+    get_grouping_list = function(Grouping,GroupingDataFrame){  #move out of public?
+      if (!is.null(GroupingDataFrame)){
+        selfGroupingDataFrame <- GroupingDataFrame
+        selfGrouping<-tail(colnames(GroupingDataFrame),1)
+      } else {
+        selfGroupingDataFrame <- NULL
+        selfGrouping <- Grouping
+      }
+      groupinglist <-list(selfGroupingDataFrame = selfGroupingDataFrame , selfGrouping = selfGrouping)
+      return(groupinglist)
+    },
+
+
     getGrouping = function(data, metaData) {
-      data$colorGrouping <- getDefaultCaptions(data = data, metaData = metaData, variableList = self$colorGrouping)
-      data$sizeGrouping <- getDefaultCaptions(data = data, metaData = metaData, variableList = self$sizeGrouping)
-      data$shapeGrouping <- getDefaultCaptions(data = data, metaData = metaData, variableList = self$shapeGrouping)
+
+      markerTypeList = list(
+        list(Grp = self$colorGrouping , Df = self$colorGroupingDataFrame , DefaultColName = "colorGrouping" ),
+        list(Grp = self$sizeGrouping  , Df = self$sizeGroupingDataFrame  , DefaultColName = "sizeGrouping"  ),
+        list(Grp = self$shapeGrouping , Df = self$shapeGroupingDataFrame , DefaultColName = "shapeGrouping" )
+      )
+
+      for ( markerType in markerTypeList){
+        if (is.null( markerType$Df )){ #case where no dataframe is supplied for grouping - use default captions
+          data[[ markerType$DefaultColName ]] <- getDefaultCaptions(data = data, metaData = metaData, variableList = markerType$Grp )
+        }
+        else { #case where dataframe is supplied for grouping
+          data <- addCaptionsColumn(data,markerType$Df,newColName = markerType$DefaultColName )
+        }
+      }
 
       return(data)
     }
   )
 )
 
+
+
+
 #' @title getDefaultCaptions
 #' @param data input data.frame with variables to group by
 #' @param metaData input data.frame with variables to group by
 #' @param variableList groups as factor levels
 #' @description
 #' createGroupingVariable create a new column that groups the grouping variable
+#' @importFrom utils tail

```

Comment:
Whenever an external function is being used, it seems that an import is required at least once in the package lifetime. Otherwise the packager complains